cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -602,43 +603,109 @@ func (m *McmManager) retry(fn func(ctx context.Context) (bool, error), resourceT
-		return nil, fmt.Errorf("Unable to fetch list of Machine objects %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Machine objects %v for machinedeployment %q", err, machinedeployment.Name)
 	}
 
 	nodeList, err := m.nodeLister.List(labels.Everything())
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Nodes %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Nodes %v", err)
 	}
 
-	var nodes []string
+	var instances []cloudprovider.Instance
 	// Bearing O(n2) complexity, assuming we will not have lot of nodes/machines, open for optimisations.
 	for _, machine := range machineList {
-		if strings.Contains(machine.Name, md.Name) {
-			var found bool
-			for _, node := range nodeList {
-				if machine.Labels["node"] == node.Name {
-					nodes = append(nodes, node.Spec.ProviderID)
-					found = true
-					break
+		instance := cloudprovider.Instance{}
+		var found bool
+		for _, node := range nodeList {
+			if machine.Labels["node"] == node.Name {
+				instance.Id = node.Spec.ProviderID
+				found = true
+				break
+			}
+		}
+		if !found {
+			// No k8s node found - either the VM is up but has not registered yet or registered but node is NotReady or MCM is unable to fulfill the request to create VM.
+			// Report a special placeholder ID so that the autoscaler can track it as an unregistered node.
+			instance.Id = placeholderInstanceIDForMachineObj(machine.Name)
+			instance.Status = &cloudprovider.InstanceStatus{
+				State:     cloudprovider.InstanceCreating,
+				ErrorInfo: getCreateErrorInfo(machine),
+			}
+		}
+		instances = append(instances, instance)
+	}
+	return instances, nil
+}
+
+func placeholderInstanceIDForMachineObj(name string) string {
+	return fmt.Sprintf("requested://%s", name)
+}
+
+// getCreateErrorInfo returns cloudprovider.InstanceErrorInfo for the machine obj
+func getCreateErrorInfo(machine *v1alpha1.Machine) *cloudprovider.InstanceErrorInfo {
+	if machine.Status.LastOperation.Type == v1alpha1.MachineOperationCreate && machine.Status.LastOperation.State == v1alpha1.MachineStateFailed && machine.Status.LastOperation.ErrorCode == machinecodes.ResourceExhausted.String() {
+		return &cloudprovider.InstanceErrorInfo{
+			ErrorClass:   cloudprovider.OutOfResourcesErrorClass,
+			ErrorCode:    machinecodes.ResourceExhausted.String(),
+			ErrorMessage: getMachineStatusErrorMessage(machine),
+		}
+	}
+	return nil
+}
+
+func getMachineStatusErrorMessage(machine *v1alpha1.Machine) string {
+	desc := machine.Status.LastOperation.Description
+	decoded, err := findCodeAndMessage(desc)
+	if err != nil {
+		return desc
+	}
+	return decoded[1]
+}
+
+// findCodeAndMessage is duplicate of https://github.com/gardener/machine-controller-manager/blob/d0fdc315087158d41f31d0c4bbbb25af9845eb0f/pkg/util/provider/machinecodes/status/status.go#L120C1-L120C1
+// TODO(himanshu-kun): update this once the function at aforementioned location is exported
+func findCodeAndMessage(encodedMsg string) ([]string, error) {
```

Comment:
Guys, we really need to fix this in future. This is non-maintenable stuff.

Reply:
Why do we need to extract the message and code? The caller uses the return value of this function to set `ErrorMessage` which could the same as encodedMsg right? You would need to extract the code to use it but that is separately available anyways.